include/tscore/DynamicStats.h

```diff
@@ -0,0 +1,153 @@
+/** @file
+
+  Dynamic Stats
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include "records/I_RecCore.h"
+#include "records/I_RecProcess.h"
+
+/**
+   @class DynamicStats
+   @details
+     tscore version of TSStat* APIs
+
+     - supported: TSStatCreate, TSStatIntIncrement/TSStatIntDecrement, TSStatFindName
+
+   CAVEAT: The librecords doesn't have APIs for "unregister stats" or "realloc rsb".
+   If you want to register ton of stats, bump -maxRecords option of traffic_manager/traffic_server.
+
+ */
+class DynamicStats
+{
+public:
+  // Modifiers
+  void init(int max_stats);
+  int create(RecT rec_type, const char *name, RecDataT data_type, RecRawStatSyncCb sync_cb, bool is_persistent = false);
+  void increment(int id, int64_t amount);
+  void set_sum(int id, int64_t sum);
+  void set_int(int id, int64_t value);
+
+  // Lookup
+  int find(const char *name) const;
+  bool is_allocated() const;
+
+private:
+  RecRawStatBlock *_rsb       = nullptr;
+  std::atomic<int> _rsb_index = 0;
+};
+
+////
+// Inline functions
+//
+inline void
+DynamicStats::init(int max_stats)
+{
+  _rsb = RecAllocateRawStatBlock(max_stats);
+}
+
+/**
+   TSStatCreate
+ */
+inline int
+DynamicStats::create(RecT rec_type, const char *name, RecDataT data_type, RecRawStatSyncCb sync_cb, bool is_persistent)
+{
+  if (name == nullptr || _rsb_index >= _rsb->max_stats) {
+    return -1;
+  }
+
+  int stat_id = _rsb_index;
+
+  if (is_persistent) {
+    int r = RecRegisterRawStat(_rsb, rec_type, name, data_type, RECP_PERSISTENT, stat_id, sync_cb);
+    if (r != REC_ERR_OKAY) {
+      return -1;
+    }
+  } else {
+    int r = RecRegisterRawStat(_rsb, rec_type, name, data_type, RECP_NON_PERSISTENT, stat_id, sync_cb);
+    if (r != REC_ERR_OKAY) {
+      return -1;
+    }
+  }
```

Comment:
Why not:
```
  int r = RecRegisterRawStat(_rsb, rec_type, name, data_type, is_persistent ? RECP_PERSISTENT : RECP_NON_PERSISTENT, stat_id, sync_cb);
  if (r != REC_ERR_OKAY) {
    return -1;
  }
```
?

Reply:
Because it's a template argument.

https://github.com/apache/trafficserver/blob/e91ca3caf92554c57c5188485682382a44bb9a69/lib/records/I_RecProcess.h#L50-L51

https://github.com/apache/trafficserver/blob/e91ca3caf92554c57c5188485682382a44bb9a69/lib/records/I_RecDefs.h#L97

Reply:
Is it better to make this function template like below?
```
  template <RecPersistT T>
  int
  create(RecT rec_type, const char *name, RecDataT data_type, RecRawStatSyncCb sync_cb)
```

Reply:
Usiing the if statement rather than a template would probably generate less object code:  https://godbolt.org/z/31oY5zMsn .

Reply:
OK, let's go with the if statement.